### PR TITLE
Persist sorting per view

### PR DESCRIPTION
**Pull request by kulmann**
## Description
This PR adds a new composable (`useRouteQueryPersisted`) which takes care of syncing non-default query params into the localStorage automatically. It has been hooked into the `useSort` and `usePagination` composables (optional, can be overwritten, e.g. with mocks for tests).

This kind of spams the localStorage a little bit, but since we want to be flexible regarding the persistence backend we probably need to be on a fine grained key-value-pair level with persisting options/settings.

With this we now persist the sort choices into the local storage per view. I.e. navigating into subfolders will keep the same sort choices. Navigating to different views will restore the historic sort choices of that respective view.

## Related Issue
- Fixes https://github.com/owncloud/web/issues/5930

## Motivation and Context
Improve UX.

## How Has This Been Tested?
Not yet, unit tests need to be added/adapted.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [ ] changelog item
- [x] fix sorting on shared with me page

### CODE REVIEW

**Comment by kulmann** [comment]
cc @dschmidt since this touches composables you have been working on

**Review by dschmidt — APPROVED** [review]
This is really nice!

When I introduced the sort composable, I was annoyed I had to make the same adjustments in several locations - this nicely gets rid of those duplications. Love it!

File: packages/web-app-files/src/composables/pagination/usePagination.ts [hunk 1]
```diff
@@ -29,6 +32,19 @@ export function usePagination<T>(options: PaginationOptions<T>): PaginationResul
...
+function createPerPageRef<T>(options: PaginationOptions<TextDecodeOptions>): ComputedRef<number> {
+  if (options.perPage) {
+    return computed(() => unref(options.perPage))
+  }
+
+  const perPageQuery = useRouteQueryPersisted({
+    name: PaginationConstants.perPageQueryName,
+    defaultValue: PaginationConstants.perPageDefault
+  })
+  return computed(() => parseInt(String(perPageQuery.value)))
```
Comment (dschmidt) [inline packages/web-app-files/src/composables/pagination/usePagination.ts:49]:
```suggestion
  return computed(() => parseInt(String(unref(perPageQuery))))
```

File: packages/web-pkg/src/composables/localStorage/useLocalStorage.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { ref, watch } from '@vue/composition-api'
+
+export const useLocalStorage = (key: string, defaultValue: any = undefined): any => {
+  const existingValue = localStorage.getItem(key)
+  const variable = ref(existingValue ? JSON.parse(existingValue) : defaultValue)
+
+  watch(
+    () => variable.value,
```
Comment (dschmidt) [inline packages/web-pkg/src/composables/localStorage/useLocalStorage.ts:8]:
```suggestion
    () => unref(variable),
```

File: packages/web-pkg/src/composables/router/useRouteQueryPersisted.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  value: QueryValue
+  source: string
+}
+
+export const useRouteQueryPersisted = (options: RouteQueryPersistedOptions): Ref<QueryValue> => {
+  const routeQueryVariable = useRouteQuery(options.name)
+  const localStorageVariable = useLocalStorage(localStorageKey(options))
+  watch(
+    (): WatcherValue => {
+      if (routeQueryVariable.value) {
```
Comment (dschmidt) [inline packages/web-pkg/src/composables/router/useRouteQueryPersisted.ts:21]:
```suggestion
      if (unref(routeQueryVariable)) {
```

Comment (dschmidt) [reply packages/web-pkg/src/composables/router/useRouteQueryPersisted.ts:21]: and following :-D

File: packages/web-app-files/src/composables/sort/useSort.ts [hunk 4]
```diff
@@ -58,6 +63,30 @@ export function useSort<T>(options: SortOptions<T>): SortResult<T> {
...
+    return ref(unref(options.sortDir))
+  }
+
+  return useRouteQueryPersisted({
+    name: options.sortDirField || SortConstants.sortDirQueryName,
+    defaultValue: firstSortableField(unref(options.fields))?.sortDir,
+    routeName: options.routeName ? unref(options.routeName) : unref(useRouteName())
+  })
+}
+
```
[the remark this answers is not in the document]
Comment (kulmann) [reply packages/web-app-files/src/composables/sort/useSort.ts:89]: I'd also be fine with 
```
routeName: unref(options.routeName || useRouteName())
```
In fact, that looks much better 😅 What do you think?

[the remark this answers is not in the document]
Comment (kulmann) [reply packages/web-app-files/src/composables/sort/useSort.ts:89]: but `unref(options.routeName)` doesn't evaluate to false, does it?

**Review by pascalwengerter — CHANGES_REQUESTED** [review]
Items per page seems to mostly work fine, just produced a `This site appears to use a scroll-linked positioning effect. This may not work well with asynchronous panning; see https://firefox-source-docs.mozilla.org/performance/scroll-linked_effects.html for further details and to join the discussion on related tools and features!` for me once in Firefox

Shared with me:
- `[Vue warn]: Property or method "targetRoute" is not defined on the instance but referenced during render.` upon page navigation with only pending shares, pending&accepted and pending&accepted&declined shares
- Accepted&declined shares table persist the same preference, but I suppose that is fine for now

Shared with others:
- Sorting by name works
- Sorting by Share receiver is broken
- Sorting by Sharedate is broken

Shared via Link
- Sorting by name works
- Sorting by Share receiver is broken
- Sorting by Sharedate is broken

**Comment by kulmann** [comment]
> Items per page seems to mostly work fine, just produced a `This site appears to use a scroll-linked positioning effect. This may not work well with asynchronous panning; see https://firefox-source-docs.mozilla.org/performance/scroll-linked_effects.html for further details and to join the discussion on related tools and features!` for me once in Firefox

Is that related to this PR? 🤔 

> Shared with me:
> 
> * `[Vue warn]: Property or method "targetRoute" is not defined on the instance but referenced during render.` upon page navigation with only pending shares, pending&accepted and pending&accepted&declined shares

That's on master, but I fixed it now. The pending shares table was referencing a targetRoute variable that didn't exist anymore - just deleted it, because you can't click on resources in the pending shares table anyway, so there is no need for a target route. Additionally fixed, that the `declined` shares table had `are-resources-clickable` set to `true`, which doesn't make sense either, so I set `are-resources-clickable` to `true` now only for accepted shares, otherwise it's `false`.


> * Accepted&declined shares table persist the same preference, but I suppose that is fine for now

Yes, unfortunately the way the table is built this is not that easily changed, so I'd like to keep it for now. The better UX would be, that accepted and declined shares have their own sorting. Would like to look into that separately.


> Shared with others:
> 
> * Sorting by name works
> * Sorting by Share receiver is broken
> * Sorting by Sharedate is broken

Sorting by `Shared with` fixed - this was again a bug from current master. The fix was to handle `undefined` values during sorting by interpreting them as empty strings for sorting (`undefined.toString()` explodes, obviously).
Sorting by sharedate was not broken, but needed a page reload after clicking the broken `Shared with` sort button.
Same for `Shared via Link`

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingExternal</strong> https://drone.owncloud.com/owncloud/web/22095/40/1
The following scenarios passed on retry:
- webUISharingExternal/federationSharing.feature:106

**Comment by ownclouders** [comment]
Results for <strong>oC10iPhone3</strong> https://drone.owncloud.com/owncloud/web/22095/49/1
The following scenarios passed on retry:
- webUISharingPublicDifferentRoles/shareByPublicLinkDifferentRoles.feature:33

**Comment by ownclouders** [comment]
Results for <strong>oC10XGAPortrait3</strong> https://drone.owncloud.com/owncloud/web/22095/45/1
The following scenarios passed on retry:
- webUISharingPublicDifferentRoles/shareByPublicLinkDifferentRoles.feature:35
- webUISharingPublicDifferentRoles/shareByPublicLinkDifferentRoles.feature:33

**Comment by ownclouders** [comment]
Results for <strong>oC10iPhone1</strong> https://drone.owncloud.com/owncloud/web/22095/47/1
The following scenarios passed on retry:
- webUIPrivateLinks/accessingPrivateLinks.feature:9

**Review by dschmidt — APPROVED** [review]

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingExternal</strong> https://drone.owncloud.com/owncloud/web/22099/40/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingPermission</strong> https://drone.owncloud.com/owncloud/web/22100/35/1
The following scenarios passed on retry:
- webUISharingPermissionsUsers/sharePermissionsUsers.feature:21

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingBasic</strong> https://drone.owncloud.com/owncloud/web/22100/53/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by kulmann** [comment]
Force merged without waiting for another CI run - the last one was green and the two commits didn't change anything that could explode in tests.

**Pull request merged**